src/txmempool.h

```diff
@@ -50,6 +50,68 @@ struct LockPoints
     LockPoints() : height(0), time(0), maxInputBlock(nullptr) { }
 };
 
+/** Epoch: RAII-style guard for using epoch-based graph traversal algorithms.
+ *     When walking ancestors or descendants, we generally want to avoid
+ * visiting the same transactions twice. Some traversal algorithms use
+ * std::set (or setEntries) to deduplicate the transaction we visit.
+ * However, use of std::set is algorithmically undesirable because it both
+ * adds an asymptotic factor of O(log n) to traverals cost and triggers O(n)
+ * more dynamic memory allocations.
+ *     In many algorithms we can replace std::set with an internal mempool
+ * counter to track the time (or, "epoch") that we began a traversal, and
+ * check + update a per-transaction epoch for each transaction we look at to
+ * determine if that transaction has not yet been visited during the current
+ * traversal's epoch.
+ *     Algorithms using std::set can be replaced on a one by one basis.
+ * Both techniques are not fundamentally incompatible across the codebase.
+ * Generally speaking, however, the remaining use of std::set for mempool
+ * traversal should be viewed as a TODO for replacement with an epoch based
+ * traversal, rather than a preference for std::set over epochs in that
+ * algorithm.
+ */
+
+class Epoch
+{
+private:
+    uint64_t m_raw_epoch;
+    bool m_guarded;
+
+public:
+    Epoch() : m_raw_epoch{0}, m_guarded{false} { }
+
+    Epoch(const Epoch&) = delete; // no copy constructor
+    Epoch& operator=(const Epoch&) = delete; // not assignable
+
+    bool guarded() const { return m_guarded; }
+
+    class Marker {
+    private:
+        uint64_t m_marker;
+
+    public:
+        Marker() : m_marker{0} { }
+
+        friend class Epoch;
+    };
+
+    class SCOPED_LOCKABLE Guard {
+    private:
+        Epoch& m_epoch;
+
+    public:
+        explicit Guard(Epoch& epoch) EXCLUSIVE_LOCK_FUNCTION(epoch) LOCKS_EXCLUDED(epoch);
```

Comment:
From the Thread Safety Analysis [docs](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquire-acquire-shared-release-release-shared):
> The caller must not hold the given capability on entry

It seems `LOCKS_EXCLUDED(epoch)` is redundant, no?